source/extensions/matching/actions/format_string/config.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "envoy/config/core/v3/substitution_format_string.pb.h"
+#include "envoy/config/core/v3/substitution_format_string.pb.validate.h"
+#include "envoy/formatter/substitution_formatter.h"
+#include "envoy/matcher/matcher.h"
+#include "envoy/server/factory_context.h"
+
+#include "source/common/matcher/matcher.h"
+#include "source/server/filter_chain_manager_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Matching {
+namespace Actions {
+namespace FormatString {
+
+class Action : public Server::FilterChainBaseAction,
```

Comment:
Would it be possible to simplify the inheritance here and avoid the diamond pattern? I think this is what causes you to change all static_casts to dynamic, right? Is it possible to make FilterChainBaseAction a template and derive it from Matcher::ActionBase?

Reply:
I've tried that. The problem is this line:
`return result->getTyped<FilterChainBaseAction>().get(filter_chains_by_name_, info);`

We'd need to say something like "\exists proto T . result->getTyped<FilterChainBaseAction<T>>` which I can't figure out how to express. Is there some construct to do that? It's almost like that we do want traits here, not a strict hierarchy.

Reply:
The only way to flatten inheritance chain here is to use the "curiously recurring" pattern. The ActionBase class needs to change to support it:

```
template <typename ProtoType, typename Base=Envoy::Matcher::Action> class ActionBase : public Base {
public:
  template <typename... Args>
  ActionBase(Args... args) : Base(args...) {}

  absl::string_view typeUrl() const override { return staticTypeUrl(); }

  static absl::string_view staticTypeUrl() {
    const static std::string typeUrl = ProtoType().GetTypeName();

    return typeUrl;
  }
};
```

The `...::FormatString::Action` just needs to derive from Server::FilterChainBaseAction.

And then you can add the helper alias `...::FormatString::ActionImpl = Matcher::ActionBase<envoy::config::core::v3::SubstitutionFormatString, ...::FormatString::Action>;`

And then in the factory instead of creating object of Action class, create ActionImpl with the same arguments.

Reply:
Done.